File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int ans=INT_MIN;
    int maxSum(TreeNode* root){
        if (!root) return 0;

        int left = max(0, maxSum(root->left));
        int right = max(0, maxSum(root->right));

        int localMax = root->val + left + right;
        ans = max(ans, localMax);

        return root->val + max(left, right);
    }
    int maxPathSum(TreeNode* root) {
        maxSum(root);
        return ans;
    }
};

File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int ans=Integer.MIN_VALUE;;
    public int dfs(TreeNode root){
        if(root==null){
            return 0;
        }

        int left=Math.max(0,dfs(root.left));
        int right=Math.max(0,dfs(root.right));

        ans=Math.max(ans,left+right+root.val);

        return root.val+Math.max(left,right);
    }
    public int maxPathSum(TreeNode root) {
        // if(root==null){
        //     return 0;
        // }
        dfs(root);
        return ans;
    }
}